server/src/main/java/org/opensearch/index/codec/fuzzy/FuzzyFilterPostingsFormat.java

```diff
@@ -0,0 +1,470 @@
+        }
+    }
+
+    class FuzzySetFieldsConsumer extends FieldsConsumer {
+        private FieldsConsumer delegateFieldsConsumer;
+        private Map<FieldInfo, FuzzySet> fuzzySets = new HashMap<>();
+        private SegmentWriteState state;
+        private List<Closeable> closeables = new ArrayList<>();
+
+        public FuzzySetFieldsConsumer(FieldsConsumer fieldsConsumer, SegmentWriteState state) {
+            this.delegateFieldsConsumer = fieldsConsumer;
+            this.state = state;
+        }
+
+        @Override
+        public void write(Fields fields, NormsProducer norms) throws IOException {
+
+            // Delegate must write first: it may have opened files
+            // on creating the class
+            // (e.g. Lucene41PostingsConsumer), and write() will
+            // close them; alternatively, if we delayed pulling
+            // the fields consumer until here, we could do it
+            // afterwards:
+            delegateFieldsConsumer.write(fields, norms);
+
+            for (String field : fields) {
+                Terms terms = fields.terms(field);
+                if (terms == null) {
+                    continue;
+                }
+                FieldInfo fieldInfo = state.fieldInfos.fieldInfo(field);
+                FuzzySet fuzzySet = fuzzySetFactory.createFuzzySet(state.segmentInfo.maxDoc(), fieldInfo.name, () -> iterator(terms));
+                if (fuzzySet == null) {
+                    break;
+                }
+                assert fuzzySets.containsKey(fieldInfo) == false;
+                closeables.add(fuzzySet);
+                fuzzySets.put(fieldInfo, fuzzySet);
+            }
+        }
+
+        private Iterator<BytesRef> iterator(Terms terms) throws IOException {
+            TermsEnum termIterator = terms.iterator();
+            return new Iterator<>() {
+
+                private BytesRef currentTerm;
+                private PostingsEnum postingsEnum;
+
+                @Override
+                public boolean hasNext() {
+                    try {
+                        do {
+                            currentTerm = termIterator.next();
+                            if (currentTerm == null) {
+                                return false;
+                            }
+                            postingsEnum = termIterator.postings(postingsEnum, 0);
+                            if (postingsEnum.nextDoc() != PostingsEnum.NO_MORE_DOCS) {
+                                return true;
+                            }
+                        } while (true);
+                    } catch (IOException ex) {
+                        throw new IllegalStateException("Cannot read terms: " + termIterator.attributes());
+                    }
+                }
+
+                @Override
+                public BytesRef next() {
+                    return currentTerm;
+                }
+            };
+        }
+
+        private boolean closed;
```

Comment:
Should this be volatile?

Reply:
Just for understanding, how is a write prevented if consumer is closed? write method doesn't check for closed. Is it happening somewhere else?

Reply:
The actual file write/flush happens only after close check is done. Since this happens on the same thread, I don't think we need volatile here.